src/fix_property_atom.cpp

```diff
@@ -26,21 +25,21 @@
 using namespace LAMMPS_NS;
 using namespace FixConst;
 
-enum{MOLECULE,CHARGE,RMASS,IVEC,DVEC,IARRAY,DARRAY};
+enum { MOLECULE, CHARGE, RMASS, IVEC, DVEC, IARRAY, DARRAY };
 
 /* ---------------------------------------------------------------------- */
 
 FixPropertyAtom::FixPropertyAtom(LAMMPS *lmp, int narg, char **arg) :
-  Fix(lmp, narg, arg),
-  nvalue(0), styles(nullptr), index(nullptr), astyle(nullptr)
+    Fix(lmp, narg, arg), nvalue(0), styles(nullptr), index(nullptr), astyle(nullptr)
 {
-  if (narg < 4) error->all(FLERR,"Illegal fix property/atom command");
+  if (narg < 4) error->all(FLERR, "Illegal fix property/atom command");
 
   restart_peratom = 1;
   wd_section = 1;
+  create_attribute = 1;
```

Comment:
@jtclemm @sjplimp this setting here and the set_arrays() function below create a problem: reading custom properties via fix property/atom from either data or restart files is broken. Instead they are reset to 0 *after* they have been read in. This breaks unit and regression tests.

I could work around this for reading data files with the following hack, but I could not figure out the equivalent change for restarts and I feel that this is not the right approach, anyway. @jtclemm can you provide some insight why this is needed for the BPM package?

I am also attaching so a small input to test this by writing out a data and restart file and then reading them back and writing out another data file. The second and third data file written has all custom properties set to 0.
[in.property.txt](https://github.com/lammps/lammps/files/8480773/in.property.txt)

```
diff --git a/src/fix_property_atom.cpp b/src/fix_property_atom.cpp
index 7be3d8f343..b4b87932b3 100644
--- a/src/fix_property_atom.cpp
+++ b/src/fix_property_atom.cpp
@@ -51,6 +51,8 @@ FixPropertyAtom::FixPropertyAtom(LAMMPS *lmp, int narg, char **arg) :
   nvalue = 0;
   values_peratom = 0;
 
+  noinit_flag = 0;
+
   while (iarg < narg) {
     if (strcmp(arg[iarg], "mol") == 0) {
       if (atom->molecule_flag)
@@ -244,6 +246,8 @@ void FixPropertyAtom::init()
 
   if (strcmp(astyle, atom->atom_style) != 0)
     error->all(FLERR, "Atom style was redefined after using fix property/atom");
+
+  noinit_flag = 0;
 }
 
 /* ----------------------------------------------------------------------
@@ -264,6 +268,8 @@ void FixPropertyAtom::read_data_section(char *keyword, int n, char *buf, tagint
     atom->map_set();
   }
 
+  noinit_flag = 1;
+
   // loop over lines of atom info
   // tokenize the line into values
   // if I own atom tag, assign its values
@@ -579,6 +585,7 @@ void FixPropertyAtom::copy_arrays(int i, int j, int /*delflag*/)
 void FixPropertyAtom::set_arrays(int i)
 {
   int k, ncol;
+  if (noinit_flag) return;
 
   for (int nv = 0; nv < nvalue; nv++) {
     if (styles[nv] == IVEC)
diff --git a/src/fix_property_atom.h b/src/fix_property_atom.h
index 5e57ef9f91..92eb970ff5 100644
--- a/src/fix_property_atom.h
+++ b/src/fix_property_atom.h
@@ -62,6 +62,7 @@ class FixPropertyAtom : public Fix {
 
   int values_peratom;    // # of values per atom, including multiple for arrays
   int nmax_old;          // length of peratom arrays the last time they grew
+  int noinit_flag;       // do not initialize added data
 };
 
 }    // namespace LAMMPS_NS


```

Reply:
> I could work around this for reading data files with the following hack, but I could not figure out the equivalent change for restarts and I feel that this is not the right approach, anyway. @jtclemm can you provide some insight why this is needed for the BPM package?

This change allowed bonded molecules to be created mid way through a simulation via fix deposit, fix pour, etc. By setting arrays to a predictable value, namely zero, fix bond/history can identify when histories need to be initialized. An older solution was to alternatively clean up data left behind when atoms moved.

Reply:
> By setting arrays to a predictable value, namely zero, fix bond/history can identify when histories need to be initialized.

can you provide or point me to a sufficiently small representative example demonstrating the flow of control and what property is provided through fix property/atom? I have some thoughts about how this issue could be addressed differently and the broken functionality restored.

Reply:
Please see commit 0f1097d7e60a49328cf582c6334cbd67b20f87e0 for an alternate approach to achieve initialization of custom properties when creating atoms from molecule files. This seems a much more logical place to do this. Beyond setting those to zero, it may also be possible to have custom properties stored in molecule files and then look them up and initialize them not just to zero.

Reply:
Here's an [example](https://github.com/jtclemm/lammps/blob/BPM/examples/bpm/pour/in.bpm.pour) of pouring molecules where bonds store their initial length and the unit vector orienting the bond. As falling atoms exit a processor domain and leave old data in fix property/atom, newly created bonds would have the wrong reference state, calculate large forces/torques, and explode.

If it would be helpful, I can reduce this script and try to find a processor layout that quickly triggers this error if the set_array() method is removed.

Reply:
Not seeing the logic of this.  Fix property/atom is where the memory is owned.  So having it zero it
out when an atom is added make sense to me.  There are other fixes which own per-atom data which do the same thing
with the same function.
Also, there are other uses for fix property/atom besides data for molecule atoms.  So putting the zeroing in atom->add_molecule_atom() doesn't seem sufficient.

Reply:
also, not sure what this means:
>> reading custom properties via fix property/atom from either data or restart files is broken
I'm using this in the AMOEBA PR, it works fine, at least from data files (haven't tried it from restarts)

Reply:
> also, not sure what this means:
 
> > > reading custom properties via fix property/atom from either data or restart files is broken

it means that the changes in this pull request which *introduce* the FixPropertyAtom::set_arrays() function and enable it break both read_restart and read_data and multiple tests depending on that functionality.

> > > I'm using this in the AMOEBA PR, it works fine, at least from data files (haven't tried it from restarts)

because *your* branch does not have the change that breaks it.

Reply:
> If it would be helpful, I can reduce this script and try to find a processor layout that quickly triggers this error if the set_array() method is removed.

Yes, that would be needed, because right now I am not seeing any issues when running with 1 or 4 processors on my (tiny) desktop.

Reply:
But if I understand, your commit removes set_arrays() from fix property/atom which means you will be breaking the zeroing for all the other use cases of fix property/atom ?

Reply:
> But if I understand, your commit removes set_arrays() from fix property/atom which means you will be breaking the zeroing for all the other use cases of fix property/atom ?

Have a look at the file in *your* checkout. Does it have a set_arrays() function?

Reply:
> Not seeing the logic of this. Fix property/atom is where the memory is owned.

I disagree. The custom properties are "owned" by the Atom class. Fix property/atom only "manages" them.
Please see the `Atom::add_custom()` function.

Reply:
@jtclemm I have enabled hyper-threading and adjusted the processor distribution and I don't see any of the problems you mention when running the bpm.pour example. There is plenty of transitions/handovers between subdomains happening (indicated by the yellow lines). https://youtube.com/shorts/vzj9ZUVUknw

Reply:
> @jtclemm I have enabled hyper-threading and adjusted the processor distribution and I don't see any of the problems you mention when running the bpm.pour example. There is plenty of transitions/handovers between subdomains happening (indicated by the yellow lines). https://youtube.com/shorts/vzj9ZUVUknw

Thank you again for the thorough testing and for cleaning up all the issues in this PR. I also ran a bunch of tests this morning/afternoon for different pours with the new commits and did not see any issues. I have a few more jobs to test, but even if there is an issue I think it's going to be a rare edge case.